zone/client.cpp

```diff
@@ -8213,33 +8229,92 @@ void Client::CashReward(uint32 copper, uint32 silver, uint32 gold, uint32 platin
 	QueuePacket(outapp.get());
 }
 
-void Client::RewardFaction(int id, int amount)
+void Client::RewardFaction(int faction_id, int amount)
 {
-	// first we hit the primary faction, even without any associations
-	SetFactionLevel2(CharacterID(), id, GetClass(), GetBaseRace(), GetDeity(), amount, false);
+	SetFactionLevel2(CharacterID(), faction_id, GetClass(), GetBaseRace(), GetDeity(), amount, false);
 
-	auto faction_assoc = content_db.GetFactionAssociationHit(id);
-	// We could log here, but since it's actually expected for some not to have entries, it would be noisy.
-	if (!faction_assoc) {
+	auto f = zone->GetFactionAssociation(faction_id);
+	if (!f) {
 		return;
 	}
 
-	// now hit them in order
-	for (int i = 0; i < MAX_FACTION_ASSOC; ++i) {
-		if (faction_assoc->hits[i].id <= 0) // we don't allow later entries
-			break;
-		if (faction_assoc->hits[i].multiplier == 0.0f) {
-			LogFaction("Bad association multiplier for ID {} entry {}", id, i + 1);
-			continue;
-		}
+	std::vector<int> faction_ids = {
```

Comment:
This section is absurdly less readable than the previous code.  

My suggestion is to keep the loop and have a convivence function that returns the correct structure field by index instead.

Reply:
Can we hold on changes to this until it's merged and retested?  Ive got all the spots I can think of tested and fixes in a [PR ](https://github.com/EQEmu/Server/pull/4039)waiting to be merged, and that includes changes to this function.  I'd like to get this sub-PR merged and then pull the whole repo down so I can test in on my entire server for a day or two before any releases are made.

Reply:
I think in the future the answer should be "no", I doubt anyone will go back and make the changes now that the merge is complete.

Reply:
It is not a big enough deal to revert it or anything so marking as resolved. It's annoying but NBD.